### PR TITLE
new impl for MemStore ; refactor of collectGarbage ; tests for MemStore, LDBStore and collectGarbage

### PR DESCRIPTION
**Pull request by nonsense**
This PR is addressing https://github.com/ethersphere/go-ethereum/issues/362 and https://github.com/ethersphere/go-ethereum/issues/342 and https://github.com/ethersphere/go-ethereum/issues/357

It introduces the following changes:
1. Rename `state.MemStore` to `state.InmemoryStore`, so that we don't confuse it with `storage.MemStore`.
2. Add simple tests for the `storage.MemStore` and `storage.LDBStore` as they are used together - simple put/get of random chunks.
3. Refactors `collectGarbage` in `LDBStore` as it wasn't working and was panicking.
4. Adds tests for `collectGarbage`
5. Introduce a new `MemStore` implementation, which uses a LRUcache from github.com/hashicorp/golang-lru

TODO:
- [x] understand LDBStore's `collectGarbage`, re-enable it if necessary and add tests
- [x] understand LDBStore's `capacity` - what should happen when we reach it? add tests
- [x] maybe fix the previous `MemStore` impl. now that we have it reproduced in test? I find it very hard to do that myself as the impl. is not documented.
- [ ] carefully review the removed `gcPos` and `gcStartPos`

### CODE REVIEW

**Comment by nonsense** [comment]
Not sure if this is an actual bug, but it looks like it. A panic that was hit while building those tests:

```
$ go test ./swarm/storage -run TestMemStoreAndLDBStore -v
=== RUN   TestMemStoreAndLDBStore
panic: runtime error: index out of range

goroutine 23 [running]:
github.com/ethereum/go-ethereum/swarm/storage.gcListPartition(0xc4202ac000, 0x2710, 0x2710, 0x0, 0xffffffffffffffff, 0x0, 0xc43fea0e70)
        /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:261 +0x166
github.com/ethereum/go-ethereum/swarm/storage.gcListSelect(0xc4202ac000, 0x2710, 0x2710, 0x0, 0xffffffffffffffff, 0x0, 0x419ad31)
        /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:283 +0x7a
github.com/ethereum/go-ethereum/swarm/storage.(*LDBStore).collectGarbage(0xc4202d4000, 0x3dcccccd)
        /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:338 +0x3f9
github.com/ethereum/go-ethereum/swarm/storage.(*LDBStore).writeBatches(0xc4202d4000)
        /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:624 +0x377
created by github.com/ethereum/go-ethereum/swarm/storage.NewLDBStore
        /Users/nonsense/code/src/github.com/ethereum/go-ethereum/swarm/storage/ldbstore.go:116 +0xfa
FAIL    github.com/ethereum/go-ethereum/swarm/storage   44.555s
```

This happens when we try to push more chunks than the capacity for Swarm for LevelDB.

File: swarm/storage/memstore_lrucache.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func NewMemStore(_ *LDBStore, capacity uint) (m *MemStore) {
+	if capacity == 0 {
+		return &MemStore{
+			disabled: true,
+		}
+	}
+
+	onEvicted := func(key interface{}, value interface{}) {
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: I agree with that. As I have said before I think we should split requests for chunks, and actual chunks in different containers, until we have a very good test suite to cover and lock behaviour for these edge cases. At the moment it is difficult to reason about those edge cases and whether they are even practical.

[the remark this answers is not in the document]
Comment (janos) [reply swarm/storage/memstore_lrucache.go]: :+1:

**Comment by nonsense** [comment]
@zelig @janos @gbalint we should have a discussion regarding the TODOs. At present I think this code is not making the codebase any worse than the current version. Please review and let me know if you think we should address something now, or continue with the rest of the failures.

File: swarm/storage/common_test.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ import (
 )
 
 var (
-	loglevel = flag.Int("loglevel", 2, "verbosity of logs")
+	loglevel = flag.Int("loglevel", 3, "verbosity of logs")
```
Comment (gbalint) [inline swarm/storage/common_test.go:37]: Minor thing, but I prefer the loglevel to be a command line parameter than a hardcoded value

**Review by gbalint — COMMENTED** [review]
I'm not against these changes, but I would really like to see @zelig 's approval as he has the most knowledge about current memstore implementation.
I also think we need some benchmarking to make sure this new implementation won't cause performace regressions for our use case.

File: swarm/storage/common_test.go (hunk 2 above)
Comment (nonsense) [reply swarm/storage/common_test.go:37]: This is what a `flag` is - it is a cmdline parameter.

Comment (gbalint) [reply swarm/storage/common_test.go:37]: Oh sorry, I did not read it well :)

**Comment by nonsense** [comment]
I am all up for benchmarking, but before we benchmark implementations, we need to make sure they are correct. Considering that our current version is panicking and I don't fully understand the different if-blocks in `MemStore`, I'd rather settle for a somewhat simpler implementation.

@zelig we should have a pair-programming session, where we review this very soon! I am available at all times, so just ping me.

File: swarm/storage/memstore_lrucache_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		Key:        make([]byte, 32),
+		ReqC:       nil,
+		SData:      make([]byte, chunkSize),
+		dbStoredC:  make(chan bool),
+		dbStoredMu: &sync.Mutex{},
+	}
+
+	rand.Read(c.SData)
+
+	binary.LittleEndian.PutUint64(c.SData[:8], chunkSize)
```
Comment (zelig) [inline swarm/storage/memstore_lrucache_test.go]: this is incorrect. SData is 8 bytes for length + 4096

Comment (nonsense) [reply swarm/storage/memstore_lrucache_test.go]: This is exactly what this is doing, no? We set `c.SData` to random bytes, and then plug the `chunkSize` to `c.SData[:8]`.

Comment (nonsense) [reply swarm/storage/memstore_lrucache_test.go]: Oh, I see what you mean. `SData` should be initialised to `make([]byte, chunkSize+8)` ?

Comment (nonsense) [reply swarm/storage/memstore_lrucache_test.go]: @gbalint you might want to fix the `FakeChunk` function in `types.go`

**Comment by nonsense** [comment]
@zelig @janos @gbalint this is ready for another review.

We should discuss the purpose of `gcPos` and `gcStartPos`.

File: swarm/storage/ldbstore.go [hunk 4]
```diff
@@ -255,98 +235,49 @@ func decodeOldData(data []byte, chunk *Chunk) {
...
+	defer it.Release()
 
-	for (gcnt < gcArraySize) && (uint64(gcnt) < s.entryCnt) {
+	garbage := []*gcItem{}
+	gcnt := 0
 
-		if (s.gcPos == nil) || (s.gcPos[0] != kpIndex) {
-			it.Seek(s.gcStartPos)
-			if it.Valid() {
-				s.gcPos = it.Key()
```
Comment (nonsense) [inline swarm/storage/ldbstore.go:329]: FYI this is one of the bugs in the garbage collector - `it.Key()` gets modified on each `it.Next()` call, and so is `s.gcPos` - therefore the `idxKey` is a single entry in the `gcArray` and we would remove at most 1 entry per call to `collectGarbage()`

File: swarm/storage/ldbstore.go [hunk 5]
```diff
@@ -47,15 +48,8 @@ var (
 )
 
 const (
-	defaultDbCapacity = 5000000
```
Comment (nonsense) [inline swarm/storage/ldbstore.go:50]: Looks like we had two params for the LDBCapacity, but this one is 100x the `defaultLDBCapacity`. I think this it too much, because 5000000 * 4kb == 20gb.

For now I leave `defaultLDBCapacity` to be 50000*4kb == 200mb, but probably we should increase it by 10x, as a default value.

Comment (zelig) [reply swarm/storage/ldbstore.go:50]: 20g is fine

File: swarm/storage/ldbstore.go (hunk 4 above)
Comment (zelig) [reply swarm/storage/ldbstore.go:329]: genious

File: swarm/storage/ldbstore.go [hunk 6]
```diff
@@ -255,98 +235,49 @@ func decodeOldData(data []byte, chunk *Chunk) {
 	chunk.Size = int64(binary.BigEndian.Uint64(data[0:8]))
 }
 
-func gcListPartition(list []*gcItem, left int, right int, pivotIndex int) int {
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/storage/ldbstore.go:277]: 👍

File: swarm/storage/dpa.go [hunk 7]
```diff
@@ -35,8 +35,9 @@ implementation for storage or retrieval.
 */
 
 const (
-	singletonSwarmDbCapacity    = 50000
-	singletonSwarmCacheCapacity = 500
+	defaultLDBCapacity                = 50000 // capacity for LevelDB
```
Comment (zelig) [inline swarm/storage/dpa.go]: 5M

**Review by zelig — APPROVED** [review]

File: swarm/storage/memstore_lrucache_test.go (hunk 3 above)
Comment (gbalint) [reply swarm/storage/memstore_lrucache_test.go]: @nonsense good point

File: swarm/storage/memstore_lrucache.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+	// it is not a request
+	c, ok := m.cache.Get(string(key))
+	if !ok {
+		return nil, ErrChunkNotFound
+	}
+	chunk := c.(*Chunk)
+	if !bytes.Equal(chunk.Key, key) {
+		panic(fmt.Errorf("MemStore.Get: chunk key %s != req key %s", chunk.Key.Hex(), key.Hex()))
+	}
```
Comment (gbalint) [inline swarm/storage/memstore_lrucache.go]: I'm not sure why we need this check, but if we need it, don't we need it in the case when the chunk is retrieved from the request cache? Becuase in that case this check does not run, we return earlier.

File: swarm/storage/memstore_lrucache.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func (m *MemStore) Get(key Key) (*Chunk, error) {
+	if m.disabled {
+		return nil, ErrChunkNotFound
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
```
Comment (gbalint) [inline swarm/storage/memstore_lrucache.go]: Although the Get function of the LRU Cache can also change the contents of the cache, it uses a Lock itself, it is thread safe. So in our code I think it is enough to RLock here. What could that cause? Parallely running requests.Get and cache.Get (because they are different cache objects), but I just don't see what problems could that cause.

**Review by gbalint — CHANGES_REQUESTED** [review]
See my few minor comments, otherwise looks really clear.

**Comment by gbalint** [comment]
Please rebase and resolve conflicts se we can see if travis is green

**Comment by nonsense** [comment]
@gbalint there were a lot of conflicts, but I think they are resolved now. Unfortunately I think we got the wrong abstraction for the `StoreParams` - for example LDBStore should not care about cache size, etc.

File: swarm/storage/memstore_lrucache.go (hunk 8 above)
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: I was wondering the same. I am not sure in what scenario this is even possible, I kept it as it was already here.

Comment (gbalint) [reply swarm/storage/memstore_lrucache.go]: I would just remove this, it is a panic anyway which should never happen. I guess this was added when somebody debugged an issue. But I'm ok with keeping it too, not a big deal

File: swarm/storage/ldbstore_test.go [hunk 10]
```diff
@@ -271,3 +274,245 @@ func BenchmarkMockDbStoreGet_1_5k(b *testing.B) {
...
+
+	for i := 0; i < n; i++ {
+		ldb.Put(chunks[i])
+	}
+
+	// wait for all chunks to be stored before continuing
+	for i := 0; i < n; i++ {
+		<-chunks[i].dbStoredC
+	}
+
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/ldbstore_test.go:472]: I am just using the API of the LDBStore, and this is fairly straight-forward, so I am not sure extracting it would have any value, it'd just add indirection.

`Put` it asynchronous, and the test expects it to be synchronous. I think we just need a better interface on the LDBStore:
1. a way to simply do a synchronous put, and know that the entry is saved, once the method returns.
2. a way to asynchronously put an entry and just continue
3. synchronous batch put of N chunks
4. asynchronous batch put of N chunks

We shouldn't have to define those 4 use-cases in every test IMO. I remember reading a PR just the other day from @nolash where he extracted that in a test, and now I'd have to duplicate that.

What do you think?

[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/ldbstore_test.go:472]: See:

1. swarm/storage/common_test.go ; line 61
```
func mputChunks(store ChunkStore, processors int, n int, chunksize int, hash hash.Hash) (hs []Key) {
```

2. swarm/storage/localstore_test.go ; line 43
```
func putChunks(store *LocalStore, chunks ...*Chunk) {
```

I don't want to introduce yet a 3rd way to do that. I suggest we refactor the interface and the tests in another PR.

[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/ldbstore_test.go:472]: Added an issue so that we don't postpone this forever - https://github.com/ethersphere/go-ethereum/issues/391

File: swarm/storage/memstore_lrucache.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+type MemStore struct {
+	cache    *lru.Cache
+	requests *lru.Cache
+	mu       sync.Mutex
+	disabled bool
+}
+
+func NewMemStore(_ *LDBStore, cacheCapacity uint, requestsCapacity uint) (m *MemStore) {
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: We have a comment next to the default value:
```
	defaultChunkRequestsCacheCapacity = 5000000 // capacity for container holding outgoing requests for chunks. should be set to LevelDB capacity
```

If one is setting the current `StoreParams` they should at least check the default values, and know what these parameters mean.

Then again I agree this is very ugly, and should be fixed very soon.

[the remark this answers is not in the document]
Comment (gbalint) [reply swarm/storage/memstore_lrucache.go]: NewMemStore by itself is an exported function, imho this worth an extra line of comment here, you don't necessarily use the default values. But this is not a blocker for sure.

[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: Will add a comment. I really feel uneasy about this whole setup, so I am not sure how to document it in a meaningful way. Nevertheless will add a comment so that at least future coders have some intuition about it.

File: swarm/storage/memstore_lrucache.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+	onEvicted := func(key interface{}, value interface{}) {
+		v := value.(*Chunk)
+		<-v.dbStoredC
+	}
+	c, err := lru.NewWithEvict(int(cacheCapacity), onEvicted)
+	if err != nil {
+		panic(err)
+	}
+
+	r, err := lru.New(int(requestsCapacity))
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: Good idea about adding an error log. I wouldn't panic as we don't want to crash the process. Requests should not stay unhandled for so long, which would be a symptom that our whole architecture around this container is not correct.

File: swarm/storage/memstore_lrucache.go (hunk 8 above)
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: Removed.

File: swarm/storage/memstore_lrucache.go (hunk 9 above)
Comment (nonsense) [reply swarm/storage/memstore_lrucache.go]: Not sure what you are proposing here? I am also not fully confident about the relationship between the two caches.

Comment (gbalint) [reply swarm/storage/memstore_lrucache.go]: I propose just to RLock, and not Lock in the Get function
